### PR TITLE
feat(learning): add lesson cooldown guard

### LINKED ISSUE #1863: feat(learning): add learning cooldown to prevent churn
## Description
Implement an LLM-friendly, independently shippable backlog item for **add learning cooldown to prevent churn**. This issue should stay narrow enough for one agent/PR while producing durable behavior, tests, and documentation where appropriate.

## Affected locations
- Lesson capture, scoring, promotion, and retirement flows
- Worker retrospectives and PM handoff summaries
- Learning telemetry and feedback surfaces

## Impact
Without this work, future agents and operators will keep rediscovering the same gap, creating avoidable coordination cost, reliability risk, or ambiguous handoffs.

## Recommendation
Implement the scoped behavior with deterministic tests, concise docs, and structured output that PM/liveness tooling can consume.

## Goals
Deliver add learning cooldown to prevent churn as a discrete capability or hardening change.
Make the behavior deterministic enough for autonomous workers to verify.
Keep the implementation small enough for a one-issue worker and one PR.
Capture operator/user-facing behavior in docs or help text when applicable.

## Expected outcomes
add learning cooldown to prevent churn is available, documented, and test-backed.
Future PM/worker handoffs can point at this issue without needing extra interpretation.
Failure modes are explicit and produce actionable messages instead of silent drift.
The change does not broaden into adjacent backlog work or unrelated refactors.

## Suggested implementation slices
Inventory the current code/docs paths that own this behavior.
Add or update the narrow model/config/schema needed for the feature or guardrail.
Implement the smallest runtime/docs change that satisfies the acceptance criteria.
Add focused fixtures/tests for success, failure, and edge cases.
Update docs/help output and handoff notes so later agents can use it.

## Acceptance criteria
- [ ] There is a dedicated implementation or documentation change for add learning cooldown to prevent churn.
- [ ] At least one automated test, fixture, or deterministic verifier covers the intended behavior.
- [ ] At least one negative/edge case is covered so the failure mode is explicit.
- [ ] Docs/help text or inline operator guidance explains how to use or interpret the behavior.
- [ ] The PR description links this issue and summarizes verification evidence.
- [ ] No unrelated production behavior is changed outside the stated scope.

## LLM handoff notes
- Treat this as a one-issue/one-PR unit.
- Start by searching for existing implementations and tests before adding new surfaces.
- Prefer structured outputs and fixtures over prose-only behavior.
- If this overlaps an existing open issue, comment with evidence and stop for PM deduplication instead of creating an unscoped implementation.
- Do not perform broad refactors or change unrelated worker/PM behavior while implementing this ticket.

## Source
Created by the 2026-07-11 fourth strategic backlog expansion run requesting 100 issues. Topic: **Learning for agents**.

### PR DESCRIPTION
**Pull request by djm204**
## Summary
- adds a deterministic learning cooldown to `LessonRecorder` so equivalent critique lessons are suppressed instead of repeatedly recorded
- exposes cooldown metadata and structured suppression results for PM/liveness tooling
- documents cooldown behavior and adds regression coverage for metadata, suppression, expiry, and invalid windows

## Verification
- `npm run test --workspace @franken/critique -- --run tests/unit/memory/lesson-recorder.test.ts`
- `npm run test --workspace @franken/critique -- --run tests/unit/types/public-contracts.test.ts tests/unit/reviewer.test.ts`
- `npm run test --workspace @franken/critique`
- `npm run typecheck --workspace @franken/critique`
- `npm run build --workspace @franken/critique`
- `npm run lint --workspace @franken/critique`
- `npx prettier --check packages/franken-critique/src/index.ts packages/franken-critique/src/memory/lesson-recorder.ts packages/franken-critique/src/types/contracts.ts packages/franken-critique/tests/unit/memory/lesson-recorder.test.ts packages/franken-critique/README.md`

Note: full package `npm run format:check --workspace @franken/critique` still reports pre-existing formatting warnings across many untouched package files; changed files pass targeted Prettier check.

Closes #1863

### CODE REVIEW

**Comment by djm204** [comment]
@codex review

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 1]
```diff
@@ -48,15 +61,38 @@ const POST_PR_LESSON_EXTRACTION_TEMPLATE: PostPrLessonExtractionTemplate = {
...
+    if (!Number.isFinite(cooldownMs) || cooldownMs < 0) {
+      throw new RangeError(
+        'LessonRecorder cooldownMs must be a finite, non-negative number.',
+      );
+    }
 
-  constructor(memory: MemoryPort) {
     this.memory = memory;
+    this.cooldownMs = cooldownMs;
+    this.now = options.now ?? ((): Date => new Date());
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in 9fa17adf: the default clock now uses @franken/types isoNow() again, while injected clocks are normalized to ISO strings for deterministic replay compatibility.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 2]
```diff
@@ -66,13 +102,31 @@ export class LessonRecorder {
...
+            continue;
+          }
+        }
+
         try {
           await this.memory.recordLesson(lesson);
+          recordingResult.recorded += 1;
+          if (cooldownKey && this.cooldownMs > 0) {
+            this.cooldowns.set(
+              cooldownKey,
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in 9fa17adf: cooldown keys are now reserved before awaiting memory persistence and rolled back if persistence fails, preventing concurrent duplicate admission.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 3]
```diff
@@ -48,15 +61,38 @@ const POST_PR_LESSON_EXTRACTION_TEMPLATE: PostPrLessonExtractionTemplate = {
 
 export class LessonRecorder {
   private readonly memory: MemoryPort;
+  private readonly cooldownMs: number;
+  private readonly now: () => Date;
+  private readonly cooldowns = new Map<string, number>();
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in 9fa17adf: recording now prunes expired cooldown entries before admitting new lessons so long-lived recorders do not retain stale unique keys indefinitely.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 4]
```diff
@@ -91,9 +145,22 @@ export class LessonRecorder {
...
+        );
         const findingMessages = critiqueFindings.map((f) => f.message);
+        const recordedAt = this.now().toISOString();
+        const cooldownKey = createCooldownKey(
+          evalResult.evaluatorName,
+          findingMessages,
+        );
+        const suppressUntil = new Date(
+          Date.parse(recordedAt) + this.cooldownMs,
+        ).toISOString();
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:222]: Fixed in 9fa17adf: constructor validation now rejects non-finite, negative, or unsupported overlarge cooldown windows before Date arithmetic can throw during record().

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 5]
```diff
@@ -167,23 +275,58 @@ function createReviewerFeedbackCapture(
...
 }
 
+function createCooldownKey(
+  evaluatorName: string,
+  findingMessages: readonly string[],
+): string {
+  const normalizedFindings = findingMessages
+    .map((message) => message.trim())
+    .sort()
+    .join('\n');
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in 9fa17adf: cooldown keys now hash a JSON-encoded normalized finding array so multiline finding boundaries remain distinct from multiple single-line findings.

**Comment by djm204** [comment]
@codex review

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 6]
```diff
@@ -66,13 +109,40 @@ export class LessonRecorder {
...
+          const suppression = this.getCooldownSuppression(lesson, cooldownKey);
+          if (suppression) {
+            recordingResult.suppressedByCooldown.push(suppression);
+            continue;
+          }
+
+          if (this.cooldownMs > 0) {
+            previousSuppressUntilMs = this.cooldowns.get(cooldownKey);
+            reservedUntilMs = Date.parse(lesson.cooldown!.suppressUntil);
+            this.cooldowns.set(cooldownKey, reservedUntilMs);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in 03995d61: duplicate callers now wait for an in-flight admission. If the admitting persistence fails, the waiting caller retries/persists instead of returning a suppression, so transient store failures do not drop the lesson.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 7]
```diff
@@ -167,23 +299,76 @@ function createReviewerFeedbackCapture(
...
+  evaluatorName: string,
+  findingMessages: readonly string[],
+): string {
+  const normalizedFindings = JSON.stringify(
+    findingMessages.map((message) => message.trim()).sort(),
+  );
+  return [
+    'critique-lesson',
+    sanitizeLessonIdPart(evaluatorName),
+    stableHash(normalizedFindings),
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:414]: Fixed in 03995d61: cooldown key material now uses SHA-256 over the JSON-encoded normalized findings, replacing the 32-bit FNV-style digest with a collision-resistant key suffix.

**Comment by djm204** [comment]
@codex review

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 8]
```diff
@@ -66,13 +111,67 @@ export class LessonRecorder {
...
+        if (cooldownKey) {
+          const suppression = this.getCooldownSuppression(lesson, cooldownKey);
+          if (suppression) {
+            recordingResult.suppressedByCooldown.push(suppression);
+            continue;
+          }
+
+          const pendingAdmission = this.pendingAdmissions.get(cooldownKey);
+          if (pendingAdmission) {
+            const admitted = await pendingAdmission;
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in 20130ba3: waiters now loop back and re-check pending admissions/cooldown state after an admitting write fails, so only one retry can reserve the key and additional concurrent waiters suppress after that retry succeeds.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 9]
```diff
@@ -48,15 +64,44 @@ const POST_PR_LESSON_EXTRACTION_TEMPLATE: PostPrLessonExtractionTemplate = {
...
+    ) {
+      throw new RangeError(
+        'LessonRecorder cooldownMs must be a finite, non-negative number within the supported Date range.',
+      );
+    }
 
-  constructor(memory: MemoryPort) {
     this.memory = memory;
+    this.cooldownMs = cooldownMs;
+    const now = options.now ?? isoNow;
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in 20130ba3: cooldown expiry now uses an advancing wall-clock source by default while lesson timestamps continue to use deterministic isoNow(); injected clocks still drive both values in tests.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 10]
```diff
@@ -167,23 +328,71 @@ function createReviewerFeedbackCapture(
...
+function createCooldownKey(
+  evaluatorName: string,
+  findingMessages: readonly string[],
+): string {
+  const normalizedFindings = JSON.stringify(
+    findingMessages.map((message) => message.trim()).sort(),
+  );
+  return [
+    'critique-lesson',
+    sanitizeLessonIdPart(evaluatorName),
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:413]: Fixed in 20130ba3: the SHA-256 cooldown-key payload now includes the raw evaluator name plus normalized findings, so sanitized display-prefix collisions cannot suppress distinct evaluators.

**Comment by djm204** [comment]
@codex review

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 11]
```diff
@@ -48,15 +64,48 @@ const POST_PR_LESSON_EXTRACTION_TEMPLATE: PostPrLessonExtractionTemplate = {
...
+      );
+    }
 
-  constructor(memory: MemoryPort) {
     this.memory = memory;
+    this.cooldownMs = cooldownMs;
+    const now = options.now ?? isoNow;
+    this.cooldownNowMs = options.now
+      ? (): number => Date.parse(normalizeTimestamp(options.now!()))
+      : (): number => Date.now();
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in d56baac0: the default recorder clock is now an advancing Date clock, and cooldown timestamps/suppression checks derive from the same normalized recorder timestamp. Tests/replay callers can still inject now explicitly.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 12]
```diff
@@ -91,9 +195,25 @@ export class LessonRecorder {
...
+        const cooldownBaseMs = this.cooldownNowMs();
+        this.pruneExpiredCooldowns(cooldownBaseMs);
+        const cooldownKey = createCooldownKey(
+          evalResult.evaluatorName,
+          findingMessages,
+        );
+        const suppressUntil = addCooldownWindow(
+          cooldownBaseMs,
+          this.cooldownMs,
+        );
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Fixed in d56baac0: local cooldown state is now set after MemoryPort.recordLesson succeeds using a fresh recorder timestamp, so slow persistence does not consume the suppression window before admission.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 13]
```diff
@@ -48,15 +64,48 @@ const POST_PR_LESSON_EXTRACTION_TEMPLATE: PostPrLessonExtractionTemplate = {
 
 export class LessonRecorder {
   private readonly memory: MemoryPort;
+  private readonly cooldownMs: number;
+  private readonly now: () => string;
+  private readonly cooldownNowMs: () => number;
+  private readonly cooldowns = new Map<string, number>();
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts]: Addressed in d56baac0 with an explicit shared-store option: LessonRecorderOptions now accepts a reused cooldownStore Map so reviewer rebuilds in the same worker can carry cooldown state across instances; the default remains instance-local to avoid leaking unrelated pipeline/test state.

**Comment by djm204** [comment]
@codex review

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 14]
```diff
@@ -91,9 +190,24 @@ export class LessonRecorder {
...
-        const lessonId = createLessonId(taskId, evalResult.evaluatorName, failingIteration.index);
+        const resolvedIteration =
+          passingIteration?.index ?? failingIteration.index;
+        const lessonId = createLessonId(
+          taskId,
+          evalResult.evaluatorName,
+          failingIteration.index,
+        );
         const findingMessages = critiqueFindings.map((f) => f.message);
+        const recordedAt = this.now();
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:213]: Fixed in db0a13ac: lessons are now passed through an admission-timestamp refresh immediately before MemoryPort.recordLesson, and the in-memory cooldown map stores the same suppressUntil that is attached to the recorded lesson. The targeted regression asserts recordedAt/suppressUntil metadata and local suppression stay aligned. Verified with npm run test --workspace @franken/critique -- --run tests/unit/memory/lesson-recorder.test.ts, npm run typecheck --workspace @franken/critique, npm run lint --workspace @franken/critique, and npm run build --workspace @franken/critique.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 15]
```diff
@@ -66,13 +113,65 @@ export class LessonRecorder {
...
+              lesson,
+              cooldownKey,
+            );
+            if (suppression) {
+              recordingResult.suppressedByCooldown.push(suppression);
+              admittedByAnotherCall = true;
+              break;
+            }
+
+            const pendingAdmission = this.pendingAdmissions.get(cooldownKey);
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:140]: Fixed in ad3ac499/db0a13ac: pending admissions are now keyed through the shared cooldownStore via a WeakMap, so recorder instances that reuse the documented shared store also share in-flight reservations. The regression 'shares pending admissions when recorder instances reuse a cooldown store' verifies concurrent duplicate records across two recorders persist only once.

File: packages/franken-critique/src/memory/lesson-recorder.ts [hunk 16]
```diff
@@ -66,13 +113,65 @@ export class LessonRecorder {
...
       const lessons = this.extractLessons(iteration, result.iterations, taskId);
       for (const lesson of lessons) {
+        const cooldownKey = lesson.cooldown?.key;
+        let admissionSettled: ((admitted: boolean) => void) | undefined;
+        let admissionPromise: Promise<boolean> | undefined;
+        if (cooldownKey) {
+          let admittedByAnotherCall = false;
+          while (!admittedByAnotherCall) {
+            const suppression = this.getCooldownSuppression(
+              lesson,
```
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:131]: Fixed in ad3ac499/db0a13ac: cooldown admission now derives cooldownKey only when cooldownMs > 0, so a recorder configured with cooldownMs: 0 ignores stale entries in a reused cooldownStore and records normally. The regression 'honors cooldownMs 0 even when a cooldown store is reused' covers this opt-out path.

File: packages/franken-critique/src/memory/lesson-recorder.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:213]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

File: packages/franken-critique/src/memory/lesson-recorder.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:140]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

File: packages/franken-critique/src/memory/lesson-recorder.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:131]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

File: packages/franken-critique/src/memory/lesson-recorder.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:213]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

**Comment by djm204** [comment]
@codex review

File: packages/franken-critique/src/memory/lesson-recorder.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:213]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

File: packages/franken-critique/src/memory/lesson-recorder.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:140]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:140]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

File: packages/franken-critique/src/memory/lesson-recorder.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:131]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

[the remark this answers is not in the document]
Comment (djm204) [reply packages/franken-critique/src/memory/lesson-recorder.ts:131]: Addressed in the latest pushed fixes; resolving this Codex thread and requesting a fresh review.

**Comment by djm204** [comment]
@codex review

**Comment by djm204** [comment]
@codex review
